src/intersection_helper.jl

```diff
@@ -0,0 +1,54 @@
+# wrapper for block_to_dimension_indices
+function get_constrained_blocks(X::CartesianProductArray{N}, Y::HPolyhedron{N}) where {N<:Real}
+    if isbounded(Y)
+        blocks, non_empty_length = block_to_dimension_indices(X)
+    else
+        constrained_vars = constrained_dimensions(Y)
+        blocks, non_empty_length = block_to_dimension_indices(X, constrained_vars)
+    end
+    return blocks, non_empty_length
+end
+
+"""
+    get_constrained_subset(X::CartesianProductArray{N, S}, Y::HPolyhedron{N}) where {N<:Real, S<:LazySet{N}}
+
+Preprocess step for intersection between cartesian product array and H-polyhedron.
+Returns low-dimensional HPolytope subset of X in constrained dimensions,
+constrained variables and variables in corresponding blocks, original block structure
+of low-dimensional subset and list of constrained blocks.
+
+### Input
+
+- `X` -- Cartesian product array of convex sets
+- `Y` -- H-polyhedron
+
+### Output
+
+A tuples of low-dimensional set, list of constrained dimensions, original block
+structure of low-dimensional set and corresponding blocks indicies
+"""
+function get_constrained_subset(X::CartesianProductArray{N, S}, Y::HPolyhedron{N}) where {N<:Real, S<:LazySet{N}}
+
+    blocks, non_empty_length = get_constrained_blocks(X, Y)
+
+    array = Vector{S}()
+    sizehint!(array, non_empty_length)
+    vars = Vector{Int}()
+    block_structure = Vector{UnitRange{Int}}()
+
+    last_var = 1
+    for i in 1:length(blocks)
+        start_index, end_index = blocks[i]
+        block_end = last_var + end_index - start_index
+        if (start_index != -1)
+            push!(array, X.array[i])
+            append!(vars, start_index : end_index)
+            push!(block_structure, last_var:(block_end))
+        end
+        last_var += block_end + 1
```

Comment:
```suggestion
        last_var = block_end + 1
```
(`last_var` is already added to `block_end`.)